versioned/storage/telemetry/src/main/java/org/projectnessie/versioned/storage/telemetry/TelemetryPersist.java

```diff
@@ -0,0 +1,405 @@
+/*
+ * Copyright (C) 2022 Dremio
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.projectnessie.versioned.storage.telemetry;
+
+import static java.util.Arrays.stream;
+import static java.util.stream.Collectors.counting;
+import static java.util.stream.Collectors.groupingBy;
+
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import javax.annotation.Nonnull;
+import org.projectnessie.versioned.storage.common.config.StoreConfig;
+import org.projectnessie.versioned.storage.common.exceptions.ObjNotFoundException;
+import org.projectnessie.versioned.storage.common.exceptions.ObjTooLargeException;
+import org.projectnessie.versioned.storage.common.exceptions.RefAlreadyExistsException;
+import org.projectnessie.versioned.storage.common.exceptions.RefConditionFailedException;
+import org.projectnessie.versioned.storage.common.exceptions.RefNotFoundException;
+import org.projectnessie.versioned.storage.common.persist.CloseableIterator;
+import org.projectnessie.versioned.storage.common.persist.Obj;
+import org.projectnessie.versioned.storage.common.persist.ObjId;
+import org.projectnessie.versioned.storage.common.persist.ObjType;
+import org.projectnessie.versioned.storage.common.persist.Persist;
+import org.projectnessie.versioned.storage.common.persist.Reference;
+
+final class TelemetryPersist implements Persist {
+
+  final Persist persist;
+  final Function<String, Traced> tracerSupplier;
+
+  TelemetryPersist(Persist persist, Function<String, Traced> tracerSupplier) {
+    this.persist = persist;
+    this.tracerSupplier = tracerSupplier;
+  }
+
+  @SuppressWarnings("resource")
+  Traced traced(String spanName) {
+    String repo = persist.config().repositoryId();
+    Traced traced = tracerSupplier.apply(spanName);
+    return repo != null && !repo.isEmpty()
+        ? traced.attribute("repo", persist.config().repositoryId())
+        : traced;
+  }
+
+  @Override
+  @Nonnull
+  @jakarta.annotation.Nonnull
+  public Reference addReference(@Nonnull @jakarta.annotation.Nonnull Reference reference)
+      throws RefAlreadyExistsException {
+    try (Traced trace = traced("addReference")) {
+      try {
+        return persist.addReference(reference);
+      } catch (RefAlreadyExistsException e) {
+        trace.attribute("error", "already exists");
```

Comment:
I think it's worth adding the exception to the span too... for trouble-shooting

Reply:
nah - it can only happen, if the reference already exists in the `refs` table